### PR TITLE
Dynapi parameters, io re-engineering

### PR DESCRIPTION
**Pull request by litvinovg**
# What does this pull request do?
* Align dynamic API ontology class names with convention
* Convert dynamic API ontology to n3
* renamed DefaultAction into NullAction
* Re-engineering of OperationData and Converters in progress
* Use jsonPath to load data from requests
* Use json schema validation to validate input and output json.
* Resource id is injected into input json body before processing by JsonConverter
* Parameter can be declared as local. Local parameters are auto-created by api before processing input data.
* ModelComponent was removed. Instead of ModelComponent DynapiModelFactory is used to create links to Vitro|VIVO models. Links are created by API for each request it required as done for all other local parameters.
* Old data types were removed
* ParameterTypes now store SerializationType and ImplementationType, also can contain type in case of arrays.
* ImplementationType store serialisation and deserialization configs and class name to be created as a result of de-serialization.
* View classes created to sort Data in DataStore, get data in most specific types required for components
* Created Implementation configurations to store information how serialization and de-serialization should be done
* Parameter conversion is done with information stored in ImplementationConfig. ConversionMethod that should do the conversion  is created during Action initialization and stored in ImplementationType. Added tests for conversion.
* Old converters were removed
* requiredParameters were renamed to InputParameters
* providedParameters were renamed to OutputParameters
* DataStore replaced OperationData
* Data class was removed and replaced with new Data class that stores Object data and serialized value provided by converters.
* Defined parameter type JsonObjectArrayOfPairs to store all results of SPARQL Query.
* Created ModelWriter component + unit and integration tests
* Implemented content type negotiation in converter, added tests

### CODE REVIEW

**Review by chenejac — CHANGES_REQUESTED** [review]
@litvinovg I have reviewed your PR, it looks very well. I have provided a couple of comments. Moreover, the list of classes in the data.types package is a little bit strange, maybe that should be split into two packages.

**Comment by chenejac** [comment]
> * Parameter can be declared as local. Local parameters are auto-created by api before processing input data.

Is this idea for InternalParameters?

File: home/src/main/resources/rdf/tbox/filegraph/vitro-dynamic-api.n3 [hunk 1]
```diff
@@ -0,0 +1,749 @@
...
+                         ] ;
+             rdfs:range xsd:string ;
+             rdfs:label "description"@en-US .
+
+###  https://vivoweb.org/ontology/vitro-dynamic-api#isInternal
+:isInternal rdf:type owl:DatatypeProperty ,
+                     owl:FunctionalProperty ;
+            rdfs:domain :ParameterType ;
+            rdfs:range xsd:boolean ;
+            rdfs:label "is internal parameter type"@en-US
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply home/src/main/resources/rdf/tbox/filegraph/vitro-dynamic-api.n3]: fixed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/ActionPool.java [hunk 2]
```diff
@@ -18,12 +20,15 @@ public ActionPool getPool() {
...
-        return new DefaultAction();
+        return new NullAction();
     }
 
     @Override
     public Class<Action> getType() {
         return Action.class;
     }
 
+    public static ContextModelAccess getModelAccess() {
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/ActionPool.java:31]: Should be removed

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java [hunk 3]
```diff
@@ -635,27 +634,27 @@ private void buildObjectSchema(ObjectSchema objectSchema, Parameters parameters)
         }
         for (String parameterName : parameters.getNames()) {
             Parameter parameter = parameters.get(parameterName);
-            ParameterType parameterType = parameter.getType();
+            SerializationType parameterType = parameter.getType().getSerializationType();
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/DynamicAPIDocumentation.java:637]: Ok.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Action.java [hunk 4]
```diff
@@ -102,29 +104,31 @@ public void removeDeadClients() {
     public boolean hasClients() {
         return !clients.isEmpty();
     }
-
-
+    
+    public Parameters getInternalParams() {
+        return internalParams;
+    }
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Action.java:110]: Parameters not required in request, but created for each action. VIVO models is one of such examples

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 5]
```diff
@@ -17,23 +21,28 @@ public class N3Template extends Operation implements Template {
...
-	private Parameters requiredParams = new Parameters();
+	private Parameters inputParams = new Parameters();
 	private String n3TextAdditions = "";
 	private String n3TextRetractions = "";
-	private ModelComponent templateModel;
-
+	private Parameter modelParam;
 	// region @Property Setters
+	
+	@Property(uri = "https://vivoweb.org/ontology/vitro-dynamic-api#hasModel", minOccurs = 1, maxOccurs = 1)
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:30]: Yes

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 6]
```diff
@@ -17,23 +21,28 @@ public class N3Template extends Operation implements Template {
...
+	public void addInputParameter(Parameter param) {
+		inputParams.add(param);
 	}
 
-	@Property(uri = "https://vivoweb.org/ontology/vitro-dynamic-api#hasModel", minOccurs = 1, maxOccurs = 1)
-	public void setTemplateModel(ModelComponent templateModel){ this.templateModel = templateModel; }
-
 	@Property(uri = "https://vivoweb.org/ontology/vitro-dynamic-api#N3TextAdditions", minOccurs = 0, maxOccurs = 1)
-	public void setN3TextAdditions(String n3TextAdditions) {
+	public void setN3Template(String n3TextAdditions) {
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:45]: A mistake,  should be fixed.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Operation.java [hunk 7]
```diff
@@ -39,28 +40,4 @@ public boolean isOutputValid(OperationData inputOutput) {
...
-                    retVal = prefix + partOfName;
-                    break;
-                }
-            }
-        }
-
-        return retVal;
-    }
-
 }
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Operation.java:43]: Ok

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Parameter.java [hunk 8]
```diff
@@ -1,37 +1,31 @@
...
 import edu.cornell.mannlib.vitro.webapp.dynapi.components.validators.Validator;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.Data;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.conversion.InitializationException;
+import edu.cornell.mannlib.vitro.webapp.dynapi.data.types.ParameterType;
 import edu.cornell.mannlib.vitro.webapp.utils.configuration.Property;
 
 public class Parameter implements Removable {
 
-    private String name;
+	private String name;
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Parameter.java:11]: Ok

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/View.java [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package edu.cornell.mannlib.vitro.webapp.dynapi.data;
+
+public interface View {
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/View.java:3]: I don't think it is possible. I think that interface should be removed.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/SimpleDataView.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package edu.cornell.mannlib.vitro.webapp.dynapi.data;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import edu.cornell.mannlib.vitro.webapp.dynapi.components.Parameter;
+import edu.cornell.mannlib.vitro.webapp.dynapi.components.Parameters;
+
+public class SimpleDataView {
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/data/SimpleDataView.java:9]: No, interface wasn't used. Should be removed.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/validators/IsInteger.java [hunk 11]
```diff
@@ -1,30 +1,54 @@
...
+        if (data.getParam().isArray()) {
+    		List array = ArrayView.getArray(data);
+			for (Object value : array) {
+				if (!isInteger(value.toString())) {
+				    return false;
+				}
+			}
+    	} else {
+    		try {
+				if (!isInteger(data.getJsonValue())) {
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/validators/IsInteger.java:33]: Not really. It is serialized value, I will rename the method. It is also a json representation. So the validator always work with Strings.
Still we will need data specific validators to be applied only to specific data types. That should be resolved in future.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/serialization/JsonObjectSerializationType.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
+package edu.cornell.mannlib.vitro.webapp.dynapi.components.serialization;
+
+import edu.cornell.mannlib.vitro.webapp.dynapi.components.Parameter;
+import edu.cornell.mannlib.vitro.webapp.dynapi.components.Parameters;
+import edu.cornell.mannlib.vitro.webapp.utils.configuration.Property;
+
+public class JsonObjectSerializationType extends SerializationType {
```
[the remark this answers is not in the document]
Comment (litvinovg) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/serialization/JsonObjectSerializationType.java:7]: It represents tree structures. So it should be separate, but maybe renamed in future.

**Comment by litvinovg** [comment]
> > * Parameter can be declared as local. Local parameters are auto-created by api before processing input data.
> 
> Is this idea for InternalParameters?
Yes

**Comment by litvinovg** [comment]
> @litvinovg I have reviewed your PR, it looks very well. I have provided a couple of comments. Moreover, the list of classes in the data.types package is a little bit strange, maybe that should be split into two packages.

Yes. I split that files into two packages.

**Review by chenejac — APPROVED** [review]
It looks ok for me now. I still don't like the name JsonObjectSerializationType, but we can consider renaming that later.

**Pull request merged**